domains/certificates/Query.API/API/DataSyncSyncer/DataSyncSyncerWorker.cs

```diff
@@ -14,36 +18,64 @@ internal class DataSyncSyncerWorker : BackgroundService
 {
     private readonly IBus bus;
     private readonly ILogger<DataSyncSyncerWorker> logger;
-    private readonly string? gsrn;
+    private readonly List<MasterData> masterData;
+    private readonly Dictionary<string, DateTimeOffset> periodStartTimeDictionary;
+    private readonly DataSyncService dataSyncService;
 
-    public DataSyncSyncerWorker(IBus bus, MockMasterDataCollection collection, ILogger<DataSyncSyncerWorker> logger)
+    public DataSyncSyncerWorker(
+        ILogger<DataSyncSyncerWorker> logger,
+        MockMasterDataCollection collection,
+        IBus bus,
+        DataSyncService dataSyncService
+    )
     {
         this.bus = bus;
         this.logger = logger;
-        var masterData = collection.Data.FirstOrDefault();
-        gsrn = masterData?.GSRN ?? null;
+        this.dataSyncService = dataSyncService;
+
+        masterData = collection.Data.ToList();
+        periodStartTimeDictionary = collection.Data.ToList()
+            .Where(it => !string.IsNullOrWhiteSpace(it.GSRN))
+            .ToDictionary(m => m.GSRN, m => m.MeteringPointOnboardedStartDate);
     }
 
-    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    protected override async Task ExecuteAsync(CancellationToken cancellationToken)
     {
-        if (string.IsNullOrWhiteSpace(gsrn))
-        {
-            logger.LogWarning("No master data loaded. Will not produce any events");
-            return;
-        }
+        dataSyncService.SetState(periodStartTimeDictionary);
 
-        var random = new Random();
-
-        while (!stoppingToken.IsCancellationRequested)
+        while (!cancellationToken.IsCancellationRequested)
         {
-            var now = DateTimeOffset.UtcNow;
-
-            var measurement = new EnergyMeasuredIntegrationEvent(gsrn, now.AddHours(-1).ToUnixTimeSeconds(), now.ToUnixTimeSeconds(), random.NextInt64(1, 42), MeasurementQuality.Measured);
-            await bus.Publish(measurement, stoppingToken);
+            await SleepToNearestHour(cancellationToken);
 
-            logger.LogInformation("Publish EnergyMeasuredIntegrationEvent");
+            foreach (var data in masterData)
+            {
+                var measurements = await dataSyncService.FetchMeasurements(data.GSRN, data.MeteringPointOwner,
+                    cancellationToken);
 
-            await Task.Delay(TimeSpan.FromSeconds(10), stoppingToken);
+                var integrationsEvents = MapToIntegrationEvents(measurements);
+                await bus.Publish(integrationsEvents, cancellationToken);
```

Comment:
What will this do in case of no integrationsEvents? Will it do nothing? Will it throw an exception?